components/browser/menu/src/main/java/mozilla/components/browser/menu/WebExtensionBrowserMenu.kt

```diff
@@ -0,0 +1,95 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.menu
+
+import android.view.View
+import android.widget.PopupWindow
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import mozilla.components.browser.menu.item.WebExtensionBrowserMenuItem
+import mozilla.components.browser.state.selector.selectedTab
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.SessionState
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.concept.engine.webextension.BrowserAction
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+import java.util.concurrent.ConcurrentHashMap
+
+typealias WebExtensionBrowserAction = BrowserAction
+
+/**
+ * A popup menu for [WebExtensionBrowserAction] menu items.
+ */
+class WebExtensionBrowserMenu internal constructor(
+    adapter: BrowserMenuAdapter,
+    private val store: BrowserStore
+) : BrowserMenu(adapter) {
+    private var scope: CoroutineScope? = null
+
+    @kotlinx.coroutines.ExperimentalCoroutinesApi
+    override fun show(
+        anchor: View,
+        orientation: Orientation,
+        endOfMenuAlwaysVisible: Boolean,
+        onDismiss: () -> Unit
+    ): PopupWindow {
+        scope = store.flowScoped { flow ->
+            flow.ifChanged { it.selectedTab }
+                .collect { state ->
+                    getOrUpdateWebExtensionMenuItems(state, state.selectedTab)
+                    invalidate()
+                }
+        }
+
+        return super.show(
+            anchor,
+            orientation,
+            endOfMenuAlwaysVisible,
+            onDismiss
+        ).apply {
+            setOnDismissListener {
+                adapter.menu = null
+                currentPopup = null
+                scope?.cancel()
+                webExtensionBrowserActions.clear()
+                onDismiss()
+            }
+        }
+    }
+
+    companion object {
+        internal val webExtensionBrowserActions = ConcurrentHashMap<String, WebExtensionBrowserMenuItem>()
```

Comment:
This doesn't need to be a concurrent map. We're not accessing from multiple threads.